src/Polly.Core/Utils/Pipeline/DelegatingComponent.cs

```diff
@@ -0,0 +1,94 @@
+internal sealed class DelegatingComponent : PipelineComponent
+{
+    private readonly PipelineComponent _component;
+
+    public DelegatingComponent(PipelineComponent component) => _component = component;
+
+    public PipelineComponent? Next { get; set; }
+
+    public override ValueTask DisposeAsync() => default;
+
+    [ExcludeFromCodeCoverage]
+    internal override ValueTask<Outcome<TResult>> ExecuteCore<TResult, TState>(
+        Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>> callback,
+        ResilienceContext context,
+        TState state)
+    {
+#if NET6_0_OR_GREATER
+        return RuntimeFeature.IsDynamicCodeSupported ? ExecuteComponent(callback, context, state) : ExecuteComponentAot(callback, context, state);
+#else
+        return ExecuteComponent(callback, context, state);
+#endif
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static ValueTask<Outcome<TResult>> ExecuteNext<TResult, TState>(
+        PipelineComponent next,
+        Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>> callback,
+        ResilienceContext context,
+        TState state)
+    {
+        if (context.CancellationToken.IsCancellationRequested)
+        {
+            return Outcome.FromExceptionAsValueTask<TResult>(new OperationCanceledException(context.CancellationToken).TrySetStackTrace());
+        }
+
+        return next.ExecuteCore(callback, context, state);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal ValueTask<Outcome<TResult>> ExecuteComponent<TResult, TState>(
+        Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>> callback,
+        ResilienceContext context,
+        TState state)
+    {
+        return _component.ExecuteCore(
+            static (context, state) => ExecuteNext(state.Next!, state.callback, context, state.state),
+            context,
+            (Next, callback, state));
+    }
+
+#if NET6_0_OR_GREATER
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal ValueTask<Outcome<TResult>> ExecuteComponentAot<TResult, TState>(
+        Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>> callback,
+        ResilienceContext context,
+        TState state)
+    {
+        // Custom state object is used to cast the callback and state to prevent infinite
+        // generic type recursion warning IL3054 when referenced in a native AoT application.
+        // See https://github.com/App-vNext/Polly/issues/1732 for further context.
+        return _component.ExecuteCore(
+            static (context, wrapper) =>
+            {
+                var callback = (Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>>)wrapper.Callback;
+                var state = (TState)wrapper.State;
+                return ExecuteNext(wrapper.Next, callback, context, state);
+            },
+            context,
+            new StateWrapper(Next!, callback, state!));
+    }
+
+    private struct StateWrapper
```

Comment:
```suggestion
    private readonly record struct StateWrapper(...);
```

nit

Reply:
I'll re-benchmark again, but I did have this locally and one point and it seemed to make it slower than without.

Reply:
No need then, it's just nitpicking.

Reply:
I still think there's noise in the benchmarks, but it came out faster anyway, so I've adopted it for that and also the terser code. These were the numbers I got:

### Before

```
BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 7.0.403
  [Host] : .NET 7.0.13 (7.0.1323.51816), X64 RyuJIT AVX2

Job=MediumRun  Toolchain=InProcessEmitToolchain  IterationCount=15
LaunchCount=2  WarmupCount=10
```

| Method                         | Mean     | Error    | StdDev   | Allocated |
|------------------------------- |---------:|---------:|---------:|----------:|
| CompositeComponent_ExecuteCore | 47.53 ns | 1.541 ns | 2.210 ns |         - |

### With `readonly`

```
BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 7.0.403
  [Host] : .NET 7.0.13 (7.0.1323.51816), X64 RyuJIT AVX2

Job=MediumRun  Toolchain=InProcessEmitToolchain  IterationCount=15
LaunchCount=2  WarmupCount=10
```

| Method                         | Mean     | Error    | StdDev   | Allocated |
|------------------------------- |---------:|---------:|---------:|----------:|
| CompositeComponent_ExecuteCore | 43.40 ns | 0.736 ns | 1.032 ns |         - |

### With `readonly record`

```
BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 7.0.403
  [Host] : .NET 7.0.13 (7.0.1323.51816), X64 RyuJIT AVX2

Job=MediumRun  Toolchain=InProcessEmitToolchain  IterationCount=15
LaunchCount=2  WarmupCount=10
```

| Method                         | Mean     | Error    | StdDev   | Allocated |
|------------------------------- |---------:|---------:|---------:|----------:|
| CompositeComponent_ExecuteCore | 37.44 ns | 0.713 ns | 0.952 ns |         - |

Reply:
Just realised it must be noise as BDN shouldn't be executing that code path anyway...

Reply:
AoT vs. not for the three ways:

### Before

```
BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 7.0.403
  [Host] : .NET 7.0.13 (7.0.1323.51816), X64 RyuJIT AVX2

Job=MediumRun  Toolchain=InProcessEmitToolchain  IterationCount=15
LaunchCount=2  WarmupCount=10
```

| Method                              | Mean     | Error    | StdDev   | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------------------------ |---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| DelegatingComponent_ExecuteCore_Jit | 33.14 ns | 2.851 ns | 3.996 ns |  1.00 |    0.00 |      - |         - |          NA |
| DelegatingComponent_ExecuteCore_Aot | 35.31 ns | 2.358 ns | 3.456 ns |  1.08 |    0.13 | 0.0025 |      24 B |          NA |

### With `readonly`

```
BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 7.0.403
  [Host] : .NET 7.0.13 (7.0.1323.51816), X64 RyuJIT AVX2

Job=MediumRun  Toolchain=InProcessEmitToolchain  IterationCount=15
LaunchCount=2  WarmupCount=10
```

| Method                              | Mean     | Error    | StdDev   | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------------------------ |---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| DelegatingComponent_ExecuteCore_Jit | 32.06 ns | 1.823 ns | 2.672 ns |  1.00 |    0.00 |      - |         - |          NA |
| DelegatingComponent_ExecuteCore_Aot | 35.52 ns | 0.482 ns | 0.721 ns |  1.12 |    0.09 | 0.0025 |      24 B |          NA |

### With `readonly record`

```
BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 7.0.403
  [Host] : .NET 7.0.13 (7.0.1323.51816), X64 RyuJIT AVX2

Job=MediumRun  Toolchain=InProcessEmitToolchain  IterationCount=15
LaunchCount=2  WarmupCount=10
```

| Method                              | Mean     | Error    | StdDev   | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------------------------ |---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| DelegatingComponent_ExecuteCore_Jit | 31.50 ns | 1.256 ns | 1.719 ns |  1.00 |    0.00 |      - |         - |          NA |
| DelegatingComponent_ExecuteCore_Aot | 35.22 ns | 1.678 ns | 2.511 ns |  1.14 |    0.09 | 0.0025 |      24 B |          NA |